src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1338,18 +1339,26 @@ public String toString()
                '}';
     }
 
-    public StringBuilder boundStacktrace(StackTraceElement[] stacktrace, int limit)
+    public static void boundStackTrace(Throwable e, int limit, HashSet<Throwable> visited)
     {
-        if (limit >= stacktrace.length) {
-            throw new IllegalArgumentException("Invalid limit size: " + limit);
-        }
+        if (e == null)
+            return;
+
+        if (visited.contains(e))
```

Comment:
your set isn't pointer identity, it is `Object.equals + Objects.hashCode` based, which is not safe for what you are doing.